### ISSUE TITLE
alternative to jsnext:main

### ISSUE THREAD

**Issue by GGAlanSmithee** [rollup/rollup#106]
For `rollup` to take full advantage of other libraries that uses es6 modules, those libraries needs to have the `jsnext:main` attribute in their package.json. In my opinion, this is a bit awkward for the following reasons:

1. I, as the consumer of a library built with es6 modules, cannot affect if that library has the `jsnext:main` attribute or not. Sure, I can talk to the library author, but in the end it's up to him or her. (I could modify the package.json of that library, but that's seems like a hack.)
2. There seems to be some resistance to using a non-standard attribute. I can't recall where now, but I've seen authors not wanting to add non-standard attributes to their package.json for the very reason that it isn't standard.
3. The only thing that's really needed for `rollup` to take full advantage of a library is that it is built with es6 modules and that inculdes its source in the npm package.

My suggestion is for `rollup` to let me, as a consumer, specify if a package uses es6 modules or not (in addition to letting producers specify `jsnext:main`). this would work like externals / globals in the rollup bundle config, like:
```javascript
jsnext {
    "<name of import>" : "<path to library src main entry point>",
    "<name of import>" : "<path to library src main entry point>",
    ...
}
```

example bundle:
```javascript
rollup.rollup({
  entry: './src/test.js',
  external: [ 'external-lib', 'external-lib-2' ],
  jsnext: {
    'futuristic-lib' : './node_modules/futuristic/src/index.js',
    'futuristic-lib-2' : './node_modules/futuristic2/src/index.js'
  }
}).then( function ( bundle ) {
  let { code, map } = bundle.generate({
    format: 'umd',
    exports: 'named',
    moduleId: 'test',
    moduleName: 'Test',
    globals: {
      'global-lib': 'GlobalLib',
      'global-lib2': 'GlobalLib2'
    }
  });
 
  fs.writeFileSync('test.js', code + '\n//# sourceMappingURL=test.js.map');
  fs.writeFileSync('test.js.map', map.toString());
});
```

I have absolutley no insight in the technical difficulties involved in doing this, was just thinking about how to make it easier for me, using `rollup`, to consume packages that are built with es6 modules.

**Comment by Victorystick** [comment]
You make a valid point @GGAlanSmithee. It's probably not realistic to expect all library authors of ES2015 modules to add a `jsnext:main` field to their package.json files. If they're only relying on what features are standardised, Rollup shouldn't have any trouble bundling them. `jsnext` is arguably broad enough to include not only include ES2015 (which Rollup currently supports), but ES2016 (ES7), Flow and other not-yet-standardised features that are possibly available in Babel and others. It _is_ a problem.

What you're suggesting is possible, right now. Just do:

```js
const jsnext = {
    'futuristic-lib' : './node_modules/futuristic/src/index.js',
    'futuristic-lib-2' : './node_modules/futuristic2/src/index.js'
};

rollup.rollup({
  entry: './src/test.js',
  // external: [ 'external-lib', 'external-lib-2' ],
  resolveExternal: function ( id, importer ) {
    if ( !( id in jsnext ) throw new Error( 'The id "' + id + '" (imported by "' + importer + '") is not defined!' );
    return jsnext[ id ];
  }  
}).then( ... );
```
( You should only define modules aa external if you _don't_ want them to be included in the bundle. So the external array should be removed from the `rollup.rollup` call in this example. )

It could make sense to supply a module to handle this use case:
```js
const mapResolver = require( 'rollup-map-resolver' );

rollup.rollup({
  entry: './src/test.js',
  resolveExternal: mapResolver({
    'futuristic-lib' : './node_modules/futuristic/src/index.js',
    'futuristic-lib-2' : './node_modules/futuristic2/src/index.js'
  })
}).then( ... );
```

**Comment by Rich-Harris** [comment]
With that approach (overriding the resolver), the fun starts when you have dependencies of dependencies – if `futuristic-lib` depends on an external ES6 module, there's no way we, as the consumers of `futuristic-lib`, trying to bundle our library, can account for that second order dependency.

The declarative approach of having a `jsnext` field in the package.json *would* alleviate that, and probably is a more pragmatic approach to using ES6 modules than expecting module authors to add the `jsnext:main` field to their own package.json files. I'd be wary of 'claiming' `jsnext` though, it's possibly too generic. (Though I don't think using a `rollup` field would be ideal, because this is supposed to be about standards.)

Even then it inherits some of the same problems as `jsnext:main` (see https://github.com/marijnh/acorn/issues/305 for some discussion) – e.g. what happens if a module author uses custom/stage 0 Babel transformers?

In short I think that something along these lines would be an improvement, but there are still some tricky problems that need a proper community consensus, which doesn't seem to be forthcoming.

**Comment by mattdesl** [comment]
> There seems to be some resistance to using a non-standard attribute. I can't recall where now, but I've seen authors not wanting to add non-standard attributes to their package.json for the very reason that it isn't standard.

Probably for the same reason many module authors do not with to support Bower, Component, etc. It becomes another avenue they need to maintain.

I'm not sure the wording `jsnext` is very future proof. What happens when ES2015 support is widespread? Is it continually pointing to the next version of JS?

A little off topic - but what I would personally like to see is less "rollup" oriented tools around dead code elimination and ES2015 bundling. For example; browserify is a pretty opinionated tool and CLI, but it builds on much smaller and less opinionated components that are decoupled from browserify:

[insert-module-globals](https://github.com/substack/insert-module-globals), [module-deps](https://www.npmjs.com/package/module-deps), [browser-pack](https://www.npmjs.com/package/browser-pack), etc.

It would be cool to have something similar here, so that all bundlers can benefit from the work being done.

**Comment by Rich-Harris** [comment]
Aye, it's a real chicken-and-egg problem. We have to solve it somehow though, otherwise we'll be stuck with legacy formats forever. There's simply no way to offer ES6 and UMD builds of the same library simultaneously without being able to point loaders/bundlers in the right direction, and if we can't do that there's no moment at which it's possible to ditch UMD.

Unless... hmm. Maybe the answer is to have entirely separate packages. What if for every package `foo` on npm, there was an equivalent `foo-jsnext` package? `foo` would be packaged up as CommonJS or UMD, `foo-jsnext` *could* just point to `src/index.js` (or whatever), but equally could point to a generated `dist` file that used ES6 module syntax but had already applied whatever source code transformations were necessary. Both packages would have the same repo, same source code, but different build steps. Is there anything obviously wrong with that idea (other than it making `npm publish` a bit complicated)?

> A little off topic - but what I would personally like to see is less "rollup" oriented tools around dead code elimination and ES2015 bundling

I know, you're one of those 'small modules' people :grinning: I like the idea in theory, especially if it helps spread some of the ideas around tree-shaking etc, but in practice it might be a little tricky. For Rollup to work, it needs to keep track of a lot of stuff (e.g. because everything is in the same scope, it sometimes needs to deconflict identifiers, which means walking the same AST that was previously generated, with the same scope annotations, manipulating the same instance of [magic-string](https://github.com/rich-harris/magic-string), with knowledge of the target format and what eventually gets exported from the bundle) in a way that doesn't lend itself well to being broken up into discrete chunks. I fear that modularising Rollup further would *increase* complexity. Moreover, it's still a research project in some ways, and things are still in flux (e.g. we just merged https://github.com/rollup/rollup/pull/99), and breaking things apart would limit our flexibility to pursue those sorts of changes.

So I'm sympathetic to the idea but I don't think it's the right time to do it just yet.

**Comment by Victorystick** [comment]
There are probably some things that could be extracted from Rollup, but I think they're quite few. The new Scope introduced in #99 could perhaps be extracted. I'm experimenting a bit more with it in a [separate repo](victorystick/virtual-scope), since I might have use for it in another project.

Beyond that, I agree with Rich that trying to break Rollup up, would do more harm than good at this point in time. It's still early (pre-1.0) days, and I'm really excited to see where this goes. :smiley:

**Comment by mattdesl** [comment]
Yup, I totally understand reasons for "just getting it done" instead of trying to muck around with existing modules or considering how it could be modularized into less opinionated components.

It's just a shame that we now have another bundler thrown into the mix, alongside jspm, webpack, browserify, Duo, etc. The ["Emo Hellscape"](https://medium.com/@trek/last-week-i-had-a-small-meltdown-on-twitter-about-npms-future-plans-around-front-end-packaging-b424dd8d367a) rant linked in Rollup's readme is partly about the frustrations of supporting the growing number of JS tools when authoring libraries and modules.

**Comment by Rich-Harris** [comment]
Will close this as there isn't really any action to take – `jsnext:main` is handled via [rollup-plugin-npm](https://github.com/rollup/rollup-plugin-npm) (Rollup itself doesn't know anything about it, as of 0.20), and any other resolution requirements could equally be met with plugins. There's some more discussion about `jsnext:main` (and alternatives) [over here](https://github.com/jsforum/jsforum/issues/5).
